pkg/cmd/roachtest/mixed_version_decommission.go

```diff
@@ -0,0 +1,287 @@
+	return func(ctx context.Context, t *test, u *versionUpgradeTest) {
+		db := u.conn(ctx, t, from)
+		var count int
+		if err := db.QueryRow(
+			`select count(*) from crdb_internal.gossip_liveness where decommissioning = true;`).Scan(&count); err != nil {
+			t.Fatal(err)
+		}
+
+		if count != 0 {
+			t.Fatalf("expected to find 0 nodes with decommissioning=false, found %d", count)
+		}
+	}
+}
+
+// checkOneMembership checks against the `membership` column in
+// crdb_internal.gossip_liveness, asserting that only one node is marked with
+// the specified membership status. This check can be only be run against
+// servers running v20.2 and beyond.
+func checkOneMembership(from int, membership string) versionStep {
+	return func(ctx context.Context, t *test, u *versionUpgradeTest) {
+		db := u.conn(ctx, t, from)
+		var count int
+		if err := db.QueryRow(
+			`select count(*) from crdb_internal.gossip_liveness where membership = $1;`, membership).Scan(&count); err != nil {
+			t.Fatal(err)
+		}
+
+		if count != 1 {
+			t.Fatalf("expected to find 1 node with membership=%s, found %d", membership, count)
+		}
+
+		var nodeID int
+		if err := db.QueryRow(
+			`select node_id from crdb_internal.gossip_liveness where decommissioning = true;`).Scan(&nodeID); err != nil {
+			t.Fatal(err)
+		}
+		t.l.Printf("n%d membership=%s", nodeID, membership)
+	}
+}
+
+// checkAllMembership checks against the `membership` column in
+// crdb_internal.gossip_liveness, asserting that all nodes are marked with
+// the specified membership status. This check can be only be run against
+// servers running v20.2 and beyond.
+func checkAllMembership(from int, membership string) versionStep {
+	return func(ctx context.Context, t *test, u *versionUpgradeTest) {
+		db := u.conn(ctx, t, from)
+		var count int
+		if err := db.QueryRow(
+			`select count(*) from crdb_internal.gossip_liveness where membership != $1;`, membership).Scan(&count); err != nil {
+			t.Fatal(err)
+		}
+
+		if count != 0 {
+			t.Fatalf("expected to find 0 nodes with membership!=%s, found %d", membership, count)
+		}
+	}
+}
+
+// uploadVersion uploads the specified cockroach binary version on the specified
+// nodes.
+func uploadVersion(nodes nodeListOption, version string) versionStep {
+	return func(ctx context.Context, t *test, u *versionUpgradeTest) {
+		// Put the binary.
+		u.uploadVersion(ctx, t, nodes, version)
+	}
+}
+
+// startVersion starts the specified cockroach binary version on the specified
+// nodes.
+func startVersion(nodes nodeListOption, version string) versionStep {
+	return func(ctx context.Context, t *test, u *versionUpgradeTest) {
+		args := startArgs("--binary=" + cockroachBinaryPath(version))
+		u.c.Start(ctx, t, nodes, args, startArgsDontEncrypt, roachprodArgOption{"--sequential=false"})
+	}
+}
+
+// runDecommissionMixedVersions runs through randomized
+// decommission/recommission processes in mixed-version clusters.
+func runDecommissionMixedVersions(
```

Comment:
Move this to the top, as it's the best entry point for reading what this test is about.